### ISSUE TITLE
REPL gets stuck with 100% CPU after ?Function

### ISSUE THREAD

**Issue by tpapp** [JuliaLang/julia#19392]
```julia
tamas@tamas ~ % julia
               _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.5.0 (2016-09-19 18:14 UTC)
 _/ |\__'_|_|_|\__'_|  |  Official http://julialang.org/ release
|__/                   |  x86_64-pc-linux-gnu

help?> Function
search: Function function functionloc cfunction @functionloc broadcast_function broadcast!_function
```
Needs a `Ctrl-C` to interrupt, which gives this backtrace:
```
^CERROR: InterruptException:
 in push! at ./set.jl:35 [inlined]
 in _subtypes(::Module, ::DataType, ::Set{Any}, ::Set{Any}) at ./reflection.jl:169
 in _subtypes(::Module, ::DataType, ::Set{Any}, ::Set{Any}) at ./reflection.jl:171 (repeats 2 times)
 in _subtypes(::Module, ::DataType) at ./reflection.jl:163
 in subtypes(::Module, ::DataType) at ./reflection.jl:177
 in summarize(::Base.AbstractIOBuffer{Array{UInt8,1}}, ::DataType, ::Base.Docs.Binding) at ./docs/Docs.jl:358
 in summarize(::Base.Docs.Binding, ::Type{T}) at ./docs/Docs.jl:323
 in doc(::Base.Docs.Binding, ::Type{T}) at ./docs/Docs.jl:264
 in doc(::Base.Docs.Binding) at ./docs/Docs.jl:246

julia> ^C
```

**Comment by yuyichao** [comment]
I believe it was just taking a long time executing `subtypes(Function)`

**Comment by StefanKarpinski** [comment]
Yikes. While technically not a bug, this is probably worth special-casing to avoid.

**Comment by yuyichao** [comment]
(And btw it is much faster on master)

**Comment by StefanKarpinski** [comment]
Doesn't this print every function? That doesn't seem useful.

**Comment by vtjnash** [comment]
(what made it faster?)

**Comment by yuyichao** [comment]
Fixing specialization heuristics? (The second run of `subtypes(Function)` on 0.5 is also fast)

**Comment by StefanKarpinski** [comment]
It's pretty fast for me too, but my point is that this is not useful behavior for the help prompt.

**Comment by fredrikekre** [comment]
Fixed by #21649
